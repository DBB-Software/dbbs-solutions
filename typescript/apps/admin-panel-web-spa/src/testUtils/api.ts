import type { QueryStatus, FetchBaseQueryError } from '@reduxjs/toolkit/query/react'
import type { SerializedError } from '@reduxjs/toolkit'

interface Action {
  error?: FetchBaseQueryError | SerializedError
  isError: boolean
  isSuccess: boolean
  status: QueryStatus
  data?: unknown
}

export const testApiErrorResponse = (action: Action, errorMessage: string) => {
  const { status, isError, error } = action

  expect(status).toEqual('rejected')
  expect(isError).toEqual(true)
  expect(error).toEqual({
    error: `Error: ${errorMessage}`,
    status: 'FETCH_ERROR'
  })
}

export const testApiSuccessResponse = (action: Action, response: unknown) => {
  const { isSuccess, status, data } = action
  expect(status).toEqual('fulfilled')
  expect(isSuccess).toEqual(true)
  expect(data).toEqual(response)
}

export const testApiUpdateSuccessResponse = (action: Pick<Action, 'data' | 'error'>, response: unknown) => {
  const { data, error } = action
  expect(error).toBeUndefined()
  expect(data).toEqual(response)
}

export const testApiUpdateErrorResponse = (action: Pick<Action, 'data' | 'error'>, errorMessage: string) => {
  const { error, data } = action

  expect(data).toBeUndefined()
  expect(error).toEqual({
    error: `Error: ${errorMessage}`,
    status: 'FETCH_ERROR'
  })
}
