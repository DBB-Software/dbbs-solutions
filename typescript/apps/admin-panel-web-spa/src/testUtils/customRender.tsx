import React, { FC, ReactElement, ReactNode } from 'react'
import { render, type RenderResult } from '@testing-library/react'
import { FormProvider, useForm } from 'react-hook-form'

export const renderWithForm = (element: ReactElement): RenderResult => {
  const Wrapper: FC<{ children: ReactNode }> = ({ children }) => {
    const methods = useForm()
    return <FormProvider {...methods}>{children}</FormProvider>
  }
  return render(element, { wrapper: Wrapper })
}

export * from '@testing-library/react'
