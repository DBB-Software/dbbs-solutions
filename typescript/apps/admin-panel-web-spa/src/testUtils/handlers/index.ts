export * from './geo'
export * from './meetingType'
