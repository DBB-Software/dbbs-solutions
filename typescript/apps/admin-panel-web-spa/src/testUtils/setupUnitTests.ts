import '@testing-library/jest-dom'
import fetchMock from 'jest-fetch-mock'

// TODO (#2029): Remove this after flaky tests fix
jest.setTimeout(30000)

beforeEach(() => {
  fetchMock.resetMocks()
})

beforeAll(() => {
  fetchMock.enableMocks()
})

afterAll(() => {
  fetchMock.disableMocks()
})
