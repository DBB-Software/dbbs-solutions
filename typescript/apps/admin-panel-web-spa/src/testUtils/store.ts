import { configureStore } from '@reduxjs/toolkit'
import { apiSlice } from '../data-access'
import { rootReducer } from '../store/reducer'

export const setupStore = ({ overrideReducer = {}, initialState = {} } = {}) => {
  const store = configureStore({
    reducer: overrideReducer || rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({ serializableCheck: false, immutableCheck: false }).concat(apiSlice.middleware),
    preloadedState: initialState
  })

  return { store }
}
