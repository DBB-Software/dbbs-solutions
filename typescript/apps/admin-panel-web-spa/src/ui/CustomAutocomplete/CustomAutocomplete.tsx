import React, { FC, useState, useEffect } from 'react'
import { Autocomplete, TextField, CircularProgress } from '@dbbs/mui-components'
import { Control, Controller, ControllerRenderProps, FieldError, FieldValues } from 'react-hook-form'

export interface Option {
  [key: string]: string | number | boolean | null | undefined
}

export interface AutoCompleteFieldProps {
  name: string
  control?: Control<FieldValues>
  label: string
  options: Option[]
  loading?: boolean
  rules?: Record<string, unknown>
  onSearch?: (searchText: string) => void
  idKey?: string
  labelKey?: string
}

const ControlledAutocomplete: FC<{
  field: ControllerRenderProps<FieldValues, string>
  fieldState: { error?: FieldError }
  options: Option[]
  onSearch?: (searchText: string) => void
  loading?: boolean
  label: string
  idKey?: string
  labelKey?: string
}> = ({ field, fieldState, options, onSearch, loading, label, idKey = 'id', labelKey = 'fullName' }) => {
  const [inputValue, setInputValue] = useState('')
  const [selectedOption, setSelectedOption] = useState<Option | null>(null)

  useEffect(() => {
    if (field.value && !inputValue) {
      const findOption = () => options.find((opt) => opt[idKey] === field.value[idKey]) || field.value

      const defaultOption = findOption()

      setSelectedOption(defaultOption || null)
      setInputValue(defaultOption?.[labelKey] || '')
    }
  }, [field.value, options, idKey, labelKey, inputValue])

  const handleInputChange = (_: unknown, value: string, reason: string) => {
    if (reason === 'input') {
      setSelectedOption(null)
    }
    if (reason === 'reset') {
      return
    }
    setInputValue(value)
    onSearch?.(value)
  }

  const handleChange = (_: unknown, value: Option | null) => {
    setSelectedOption(value)
    field.onChange(value ? { id: value[idKey], ...value } : null)
  }

  return (
    <Autocomplete
      options={options}
      getOptionLabel={(option) => option?.[labelKey]?.toString() || ''}
      isOptionEqualToValue={(option, value) => option?.[idKey]?.toString() === value?.[idKey]}
      loading={loading}
      onInputChange={handleInputChange}
      onChange={handleChange}
      value={selectedOption}
      inputValue={inputValue}
      renderInput={(params) => (
        <TextField
          {...params}
          label={label}
          error={!!fieldState.error}
          helperText={fieldState.error?.message}
          InputProps={{
            ...params.InputProps,
            endAdornment: (
              <>
                {loading ? <CircularProgress color="inherit" size={20} /> : null}
                {params.InputProps.endAdornment}
              </>
            )
          }}
          fullWidth
        />
      )}
    />
  )
}

export const CustomAutocomplete: FC<AutoCompleteFieldProps> = ({
  name,
  control,
  label,
  options,
  loading = false,
  rules = {},
  onSearch,
  idKey = 'id',
  labelKey = 'fullName'
}) => (
  <Controller
    name={name}
    control={control}
    rules={rules}
    render={({ field, fieldState }) => (
      <ControlledAutocomplete
        field={field}
        fieldState={fieldState}
        options={options}
        loading={loading}
        label={label}
        onSearch={onSearch}
        idKey={idKey}
        labelKey={labelKey}
      />
    )}
  />
)
