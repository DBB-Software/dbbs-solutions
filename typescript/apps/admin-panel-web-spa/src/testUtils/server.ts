import { setupServer } from 'msw/node'
import { geoHandlers, meetingTypeHandlers } from './handlers'

export const mswServer = setupServer(...geoHandlers, ...meetingTypeHandlers)
