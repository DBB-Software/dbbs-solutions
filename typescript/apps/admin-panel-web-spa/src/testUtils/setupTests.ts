import '@testing-library/jest-dom'
import fetchMock from 'jest-fetch-mock'
import { mswServer } from './server'

beforeEach(() => {
  fetchMock.resetMocks()
})

beforeAll(() => {
  fetchMock.enableMocks()
  mswServer.listen()
})

afterAll(() => {
  fetchMock.disableMocks()
  mswServer.close()
})
