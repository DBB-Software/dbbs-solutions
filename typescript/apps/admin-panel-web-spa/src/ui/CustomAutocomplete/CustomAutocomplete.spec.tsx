import React from 'react'
import { CustomAutocomplete, Option } from './CustomAutocomplete'
import { renderWithForm } from '../../testUtils/customRender'

describe('<CustomAutocomplete />', () => {
  const options: Option[] = [
    { id: '1', label: 'Option 1' },
    { id: '2', label: 'Option 2' }
  ]

  it('should render correctly', () => {
    const { asFragment } = renderWithForm(<CustomAutocomplete name="test" label="Test Label" options={options} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
