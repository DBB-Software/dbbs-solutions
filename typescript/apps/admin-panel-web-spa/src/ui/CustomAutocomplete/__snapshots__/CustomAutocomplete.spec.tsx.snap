// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomAutocomplete /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-ka7ti6-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Test Label
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-p1olib-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Test Label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
